### PR TITLE
Extract PDF on file system instead of memory

### LINKED ISSUE #16955: Extract PDF on file system instead of memory
In PdfMediaFileTextProvider.cs the input stream is loaded in memory if it's not seekable. I think we may want to use the file system instead and delete the temporary file when the operation is done.

### PR DESCRIPTION
**Pull request by hishamco**
Fixes #16955

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.Media.Indexing.Pdf/Services/PdfMediaFileTextProvider.cs [hunk 1]
```diff
@@ -39,7 +38,16 @@ public async Task<string> GetTextAsync(string path, Stream fileStream)
             if (seekableStream != null)
             {
                 await seekableStream.DisposeAsync();
+
+                File.Delete(seekableStream.Name);
```
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Media.Indexing.Pdf/Services/PdfMediaFileTextProvider.cs:42]:
```suggestion
                await File.DeleteAsync(seekableStream.Name);
```

File: src/OrchardCore.Modules/OrchardCore.Media.Indexing.Pdf/Services/PdfMediaFileTextProvider.cs [hunk 2]
```diff
@@ -39,7 +38,16 @@ public async Task<string> GetTextAsync(string path, Stream fileStream)
...
             }
         }
     }
+
+    private static FileStream CreateTemporaryFile()
```
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Media.Indexing.Pdf/Services/PdfMediaFileTextProvider.cs:47]: GetTemporaryFileStream

Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Media.Indexing.Pdf/Services/PdfMediaFileTextProvider.cs:47]:
```suggestion
    private static FileStream GetTemporaryFileStream()
```

File: src/OrchardCore.Modules/OrchardCore.Media.Indexing.Pdf/Services/PdfMediaFileTextProvider.cs [hunk 3]
```diff
@@ -11,16 +11,15 @@ public async Task<string> GetTextAsync(string path, Stream fileStream)
...
+        FileStream seekableStream = null;
         try
         {
             if (!fileStream.CanSeek)
             {
-                // Since fileStream.Length might not be supported either, we can't preconfigure the capacity of the
-                // MemoryStream.
-                seekableStream = new MemoryStream();
-                // While this involves loading the file into memory, we don't really have a choice.
+                seekableStream = CreateTemporaryFile();
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Media.Indexing.Pdf/Services/PdfMediaFileTextProvider.cs:19]:
```suggestion
                seekableStream = GetTemporaryFileStream();
```

File: src/OrchardCore.Modules/OrchardCore.Media.Indexing.Pdf/Services/PdfMediaFileTextProvider.cs [hunk 4]
```diff
@@ -39,7 +38,16 @@ public async Task<string> GetTextAsync(string path, Stream fileStream)
             if (seekableStream != null)
             {
                 await seekableStream.DisposeAsync();
+
+                await File.DeleteAsync(seekableStream.Name);
```
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Media.Indexing.Pdf/Services/PdfMediaFileTextProvider.cs:42]:
```suggestion
                await Task.Run(() => File.Delete(seekableStream.Name));
```

File: src/OrchardCore.Modules/OrchardCore.Media.Indexing.Pdf/Services/PdfMediaFileTextProvider.cs [hunk 5]
```diff
@@ -39,7 +39,7 @@ public async Task<string> GetTextAsync(string path, Stream fileStream)
             {
                 await seekableStream.DisposeAsync();
 
-                await File.DeleteAsync(seekableStream.Name);
+                await Task.Run(() => File.Delete(seekableStream.Name));
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Media.Indexing.Pdf/Services/PdfMediaFileTextProvider.cs:42]: Why did you make this change, while you were requesting the `File.DeleteAsync()`

Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Media.Indexing.Pdf/Services/PdfMediaFileTextProvider.cs:42]: DeleteAsync does not exists.

File: src/OrchardCore.Modules/OrchardCore.Media.Indexing.Pdf/Services/PdfMediaFileTextProvider.cs [hunk 6]
```diff
@@ -39,7 +38,16 @@ public async Task<string> GetTextAsync(string path, Stream fileStream)
             if (seekableStream != null)
             {
                 await seekableStream.DisposeAsync();
+
+                await Task.Run(() => File.Delete(seekableStream.Name));
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Media.Indexing.Pdf/Services/PdfMediaFileTextProvider.cs:42]: No need to use `Task.Run()` here

File: src/OrchardCore.Modules/OrchardCore.Media.Indexing.Pdf/Services/PdfMediaFileTextProvider.cs [hunk 7]
```diff
@@ -11,16 +11,15 @@ public async Task<string> GetTextAsync(string path, Stream fileStream)
...
+        FileStream seekableStream = null;
         try
         {
             if (!fileStream.CanSeek)
             {
-                // Since fileStream.Length might not be supported either, we can't preconfigure the capacity of the
-                // MemoryStream.
-                seekableStream = new MemoryStream();
-                // While this involves loading the file into memory, we don't really have a choice.
+                seekableStream = new FileStream(Path.GetTempFileName(), FileMode.Create, FileAccess.Write, FileShare.None, 4096, FileOptions.DeleteOnClose | FileOptions.Asynchronous);
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Media.Indexing.Pdf/Services/PdfMediaFileTextProvider.cs:19]: @hishamco please check the differences between `GetTempFileName()` and `GetRandomFileName()`, you'll see what was wrong with your combine.

**Review by sebastienros — APPROVED** [review]

File: src/OrchardCore.Modules/OrchardCore.Media.Indexing.Pdf/Services/PdfMediaFileTextProvider.cs (hunk 7 above)
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Media.Indexing.Pdf/Services/PdfMediaFileTextProvider.cs]: I didn't use `GetRandomFileName()` :)

**Pull request merged**